### ISSUE TITLE
Suspicious `Rep (..)` import in `Cabal-syntax`'s `Distribution.Compat.Prelude`

### ISSUE THREAD

**Issue by RyanGlScott** [haskell/cabal#9070]
**Describe the bug**
`Distribution.Compat.Prelude` currently imports `GHC.Generics` in a suspicious way:

https://github.com/haskell/cabal/blob/1723d220e8de79a0d99d3127784167077fba77b4/Cabal-syntax/src/Distribution/Compat/Prelude.hs#L251

In particular, note the `Rep (..)` part. What does this mean? `Rep` is an associated type family of `Generic`, and there are no other identifiers "associated" with an associated type family. What's more, this line imports `Generic` _without_ importing any of its methods (`from` and `to`), despite the fact that `from` is used later in the module:

https://github.com/haskell/cabal/blob/1723d220e8de79a0d99d3127784167077fba77b4/Cabal-syntax/src/Distribution/Compat/Prelude.hs#L279

The fact that this compiles at all is a lucky fluke. See [GHC#23570](https://gitlab.haskell.org/ghc/ghc/-/issues/23570). What's more, it's fragile: if `GHC.Generics` were to be refactored such that it re-exports the `Generic` class from a different module, then GHC _would_ reject this code. I actually encountered this issue in practice in some in-progress GHC work that I'm doing (see [here]() if you're curious), so this line of code in `Cabal-syntax` stands in the way of that work.

**To Reproduce**

At the moment, there's no issue with any released versions of GHC (and therefore nothing to reproduce). The only way to reproduce an actual build failure would be to compile `Cabal-syntax` with the development version of GHC contained in [this in-progress merge request](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10735).

**Expected behavior**

I think the import should instead be this, which avoids any of the issues in [GHC#23570](https://gitlab.haskell.org/ghc/ghc/-/issues/23570):

```diff
diff --git a/Cabal-syntax/src/Distribution/Compat/Prelude.hs b/Cabal-syntax/src/Distribution/Compat/Prelude.hs
index 5edf0d92e..3cbf3c17a 100644
--- a/Cabal-syntax/src/Distribution/Compat/Prelude.hs
+++ b/Cabal-syntax/src/Distribution/Compat/Prelude.hs
@@ -248,7 +248,7 @@ import Data.Word (Word, Word16, Word32, Word64, Word8)
 import Distribution.Compat.Binary (Binary (..))
 import Distribution.Compat.Semigroup (Semigroup (..), gmappend, gmempty)
 import Distribution.Compat.Typeable (TypeRep, Typeable, typeRep)
-import GHC.Generics (Generic, K1 (unK1), M1 (unM1), Rep (..), U1 (U1), V1, (:*:) ((:*:)), (:+:) (L1, R1))
+import GHC.Generics (Generic (..), K1 (unK1), M1 (unM1), U1 (U1), V1, (:*:) ((:*:)), (:+:) (L1, R1))
 import System.Exit (ExitCode (..), exitFailure, exitSuccess, exitWith)
 import Text.Read (readMaybe)


```

**Comment by RyanGlScott** [comment]
I've applied my suggested fix in #9071. Let me know if anything else is required for a `Cabal-syntax` patch.

**Comment by simonpj** [comment]
> At the moment, there's no issue with any released versions of GHC (and therefore nothing to reproduce).

But the fact that there is no error is a *bug* in GHC.  One that we will fix.  And when we do, this module will start failing.  Applying Ryan's patch will continue to work with current GHCs, and with the future bug-fixed versions
